### PR TITLE
Feature/authenticate login functionality

### PR DESCRIPTION
**Pull request by HarisShahidHameed**
Trello : https://trello.com/b/FTaaANh4/be-prepared-application

### CODE REVIEW

File: server/DataBase/models/tokens.js [hunk 1]
```diff
@@ -0,0 +1,9 @@
+const mongoose = require('mongoose');
+
+let TokensSchema = new mongoose.Schema({
+    token : String,
+    createdAt: {type: Date, expires: 172800000, default: Date.now}
+});
+
```
Comment (abdulhaqshah) [inline server/DataBase/models/tokens.js:10]: extra lines

File: server/libs/userlibs.js [hunk 2]
```diff
@@ -1,51 +1,116 @@
...
+const {fieldsValidator} = require('./../utilities/utilityFunctions')
+const {statusCodes, messages} = require ('../utilities/constants');
 
-//register a new user to the app
-exports.register = function(req,res) {
-    let body = _.pick(req.body, ['name', 'email','password']);
-    let user = new User(body);
+const register = function(userData) {
+    return new Promise((resolve,reject) => {
+        if(!userData.name) {
```
Comment (abdulhaqshah) [inline server/libs/userlibs.js]: make validation function and handle all that in it

File: server/libs/userlibs.js [hunk 3]
```diff
@@ -1,51 +1,116 @@
...
+                data : user
+            })
+        }).catch((error) => {
+            if(error.code === 11000) {
+                reject({
+                    status : statusCodes.badRequest,
+                    message: `Email ${messages.duplicate}`   
+                })
+            }
+            reject({
```
Comment (abdulhaqshah) [inline server/libs/userlibs.js]: revamp this functionality

File: server/libs/userlibs.js [hunk 4]
```diff
@@ -1,51 +1,116 @@
...
+        })
+    })
+};
 
-    user.save().then(() => {
-        res.status(statusCodes.created).send({message : `User ${messages.created}`, data : user});
-    }).catch ((error) => {
-        res.status(statusCodes.bad_request).send(error);
-    }); 
+const login = function(body){
```
Comment (abdulhaqshah) [inline server/libs/userlibs.js]: spce

File: server/libs/userlibs.js [hunk 5]
```diff
@@ -1,51 +1,116 @@
...
+        User.findOne({email : body.email, deleted : false}).then((user) => {
+            if(!user) {
+                reject({
+                    status : statusCodes.notFound,
+                    message : `User ${messages.notFound}`
+                });
+            }
+            bcrypt.compare(body.password, user.password , (err,result) => {            
+                if(result) {
+                    let access = 'authentication';
```
Comment (abdulhaqshah) [inline server/libs/userlibs.js]: constant

File: server/libs/userlibs.js [hunk 6]
```diff
@@ -1,51 +1,116 @@
...
+                    let token = jwt.sign({_id: user._id.toHexString()}, access, process.env.JWT_SECRET, {expiresIn: '2d'}).toString();
+                    resolve({
+                        status : statusCodes.successful,
+                        message : `Login ${messages.successful}`,
+                        data : user,
+                        token
+                    });
+                } else {
+                    reject({
+                        status : statusCodes.forbidden,
```
Comment (abdulhaqshah) [inline server/libs/userlibs.js]: camel case

File: server/libs/userlibs.js [hunk 7]
```diff
@@ -1,51 +1,116 @@
...
-                if(res){
-                    resolve(user);
-                }else{
-                    const data = {
-                        message : `Password ${messages.not_match}`,
-                        status : statusCodes.forbidden
-                    };
-                    reject(data);
+    return new Promise((resolve,reject) => {
+        if(error){
```
Comment (abdulhaqshah) [inline server/libs/userlibs.js]: space

File: server/libs/userlibs.js [hunk 8]
```diff
@@ -1,51 +1,116 @@
...
+            });
+        })
+    })
 };
 
-//login functionality for the app. Checks whether user exist and if exist the information provided is correct
-exports.login = function(req,res){
-    let body = _.pick(req.body , ['email','password']);
+const updateUser = function (body) {
+    let error = fieldsValidator(body);
```
Comment (abdulhaqshah) [inline server/libs/userlibs.js]: nameing convention

File: server/middleware/authenticate.js [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+                    return Promise.reject({status : statusCodes.unauthorized, message : messages.unauthorized});
+                }
+                req.user = user;
+                req.token = token;
+                next();
+            }).catch((err) => {
+                res.status(statusCodes.unauthorized).send({status : statusCodes.unauthorized, message : messages.unauthorized});
+            });
+        }
+        return Promise.reject({status : statusCodes.unauthorized, message : messages.unauthorized});
```
Comment (abdulhaqshah) [inline server/middleware/authenticate.js]: do it with res on every promise

File: server/middleware/authenticate.js [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+        if(req.params.id === decoded._id) {
+            return User.findById(req.params.id);
+        }
+        return Promise.reject({status : statusCodes.unauthorized, message : messages.unauthorized});
+    }
+    if(req.body.id === decoded._id) {
+        console.log(decoded);//eslint-disable-line no-console
+        return User.findById(req.body.id);
+    }
+    return Promise.reject({status : statusCodes.unauthorized, message : messages.unauthorized});
```
Comment (abdulhaqshah) [inline server/middleware/authenticate.js]: why Promise.return

**Review by abdulhaqshah — CHANGES_REQUESTED** [review]

File: .eslintrc.json [hunk 11]
```diff
@@ -89,7 +89,7 @@
         "implicit-arrow-linebreak": "error",
         "indent": "off",
         "indent-legacy": "off",
-        "init-declarations": "error",
+        "init-declarations": "off",
```
Comment (abdulhaqshah) [inline .eslintrc.json]: revert these changes why they are off

File: server/DataBase/models/tokens.js [hunk 12]
```diff
@@ -0,0 +1,8 @@
+const mongoose = require('mongoose');
+
+let TokensSchema = new mongoose.Schema({
+    token : String,
```
Comment (abdulhaqshah) [inline server/DataBase/models/tokens.js]: make this a requried field

File: server/libs/userlibs.js [hunk 13]
```diff
@@ -1,51 +1,107 @@
...
+const {statusCodes, messages, secretKeys} = require ('../utilities/constants');
 
-//register a new user to the app
-exports.register = function(req,res) {
-    let body = _.pick(req.body, ['name', 'email','password']);
-    let user = new User(body);
+const register = function(userData) {
+    return new Promise((resolve,reject) => {
+        let result = fieldsValidator(userData);
+        if(result) {
```
Comment (abdulhaqshah) [inline server/libs/userlibs.js]: if should be like this if () {

File: server/DataBase/models/tokens.js (hunk 1 above)
Comment (HarisShahidHameed) [reply server/DataBase/models/tokens.js:10]: Removed.

File: server/libs/userlibs.js (hunk 2 above)
Comment (HarisShahidHameed) [reply server/libs/userlibs.js]: Done that.

File: server/libs/userlibs.js (hunk 3 above)
Comment (HarisShahidHameed) [reply server/libs/userlibs.js]: Done

File: server/libs/userlibs.js (hunk 4 above)
Comment (HarisShahidHameed) [reply server/libs/userlibs.js]: Given space.

File: server/libs/userlibs.js [hunk 14]
```diff
@@ -1,51 +1,107 @@
...
+const {fieldsValidator} = require('./../utilities/utilityFunctions')
+const {statusCodes, messages, secretKeys} = require ('../utilities/constants');
 
-//register a new user to the app
-exports.register = function(req,res) {
-    let body = _.pick(req.body, ['name', 'email','password']);
-    let user = new User(body);
+const register = function(userData) {
+    return new Promise((resolve,reject) => {
+        let result = fieldsValidator(userData);
```
Comment (abdulhaqshah) [inline server/libs/userlibs.js]: change this into `fieldsValidate`

File: server/libs/userlibs.js (hunk 5 above)
Comment (HarisShahidHameed) [reply server/libs/userlibs.js]: Done that. Move the secret key to constants.js and calling it from there

File: server/libs/userlibs.js (hunk 6 above)
Comment (HarisShahidHameed) [reply server/libs/userlibs.js]: Done

File: server/libs/userlibs.js (hunk 7 above)
Comment (HarisShahidHameed) [reply server/libs/userlibs.js]: Given space.

File: server/libs/userlibs.js (hunk 8 above)
Comment (HarisShahidHameed) [reply server/libs/userlibs.js]: Done.

File: server/middleware/authenticate.js (hunk 9 above)
Comment (HarisShahidHameed) [reply server/middleware/authenticate.js]: Done that.

File: server/middleware/authenticate.js (hunk 10 above)
Comment (HarisShahidHameed) [reply server/middleware/authenticate.js]: So that .catch will catch these objects and return as error.

**Review by abdulhaqshah — CHANGES_REQUESTED** [review]

File: server/middleware/authenticate.js [hunk 15]
```diff
@@ -0,0 +1,47 @@
+let User = require('./../DataBase/models/user');
+let Token = require('./../DataBase/models/tokens');
+const jwt = require('jsonwebtoken');
+const {statusCodes, messages, secretKeys} = require ('../utilities/constants');
+const {isEmpty} = require('./../utilities/utilityFunctions');
+
+const findByToken = function(token,req) {
+    let decoded;
+
```
Comment (abdulhaqshah) [inline server/middleware/authenticate.js]: remove space

File: server/middleware/authenticate.js [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+    try {
+        decoded = jwt.verify(token, secretKeys.tokenKey, process.env.JWT_SECRET);
+    } catch (error) {
+        return Promise.reject(error);
+    }
+    
+    if(!isEmpty(req.params)) {
```
Comment (abdulhaqshah) [inline server/middleware/authenticate.js]: remove

File: .eslintrc.json (hunk 11 above)
Comment (HarisShahidHameed) [reply .eslintrc.json]: Done

File: server/DataBase/models/tokens.js (hunk 12 above)
Comment (HarisShahidHameed) [reply server/DataBase/models/tokens.js]: Done

File: server/libs/userlibs.js (hunk 13 above)
Comment (HarisShahidHameed) [reply server/libs/userlibs.js]: done

File: server/libs/userlibs.js (hunk 14 above)
Comment (HarisShahidHameed) [reply server/libs/userlibs.js]: Done

File: server/middleware/authenticate.js [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+            decoded = jwt.verify(token, secretKeys.tokenKey, process.env.JWT_SECRET);
+        } catch (error) {
+            reject(error);
+        }
+        if (id === decoded._id) {
+            resolve(User.findById(req.body.id));
+        }
+        reject({status : statusCodes.unauthorized, message : messages.unauthorized});
+    })
+
```
Comment (abdulhaqshah) [inline server/middleware/authenticate.js]: extra line

**Review by abdulhaqshah — CHANGES_REQUESTED** [review]

File: server/middleware/authenticate.js (hunk 15 above)
Comment (HarisShahidHameed) [reply server/middleware/authenticate.js]: Removed.

File: server/middleware/authenticate.js (hunk 16 above)
Comment (HarisShahidHameed) [reply server/middleware/authenticate.js]: Removed.

File: server/middleware/authenticate.js (hunk 17 above)
Comment (HarisShahidHameed) [reply server/middleware/authenticate.js]: Removed

**Review by abdulhaqshah — APPROVED** [review]

**Pull request merged**